### PR TITLE
WiP: Refine Docs

### PR DESCRIPTION
**Pull request by wgao19**
Regarding issue #39 

Thoughts on how we structure the docs:

**`README.md` under root directory to contain only essential project information**
- Badges & code coverages
- Motivation
- (perhaps Quick Start?
- Contributing
- Acknowledgements
- License

**Move docs content to `documentation`, which will be the main docs directory**
I'm thinking of dividing the docs into
- **Installation**
- **Quick Start** to include a guide that covers most of the use cases
- **Syntax** contains the official syntax for the CLI itself
- **Plugins**
Guides to plugins. This will be a multi page doc section, separated by plugins.
Also provide a guide for people to build their own plugins.
- **Builders** 
I'm thinking of separating the bindings to type systems such as `TypeScript` or `FlowType` from `Plugins`. Calling it builders now (and definitely open to discussion), this part explains how builders work and provides some guides for people to build their own bindings.

**Checklist**

- [x] Add _Motivation_ section to root README.md
- [ ] Refactor and work on individual sections:
  - [x] Quick Start
  - [x] Plugins, builders and mockers
  - [ ] Syntax

### CODE REVIEW

**Comment by Cryrivers** [comment]
Lets added it!

File: README.md [hunk 1]
```diff
@@ -14,6 +13,72 @@ Contents
...
+  birthday: number; // 1529370938452
+
+  /**
+   * @example Croatia
+   */
+  country: string;  // Croatia
+}
+```
+
+### Mock error conditions specific to your type definitions
```
Comment (Cryrivers) [inline README.md]: could we remove the word `error`?

File: README.md [hunk 2]
```diff
@@ -14,6 +13,72 @@ Contents
...
+```ts
+interface User {
+  /**
+   * @faker {{internet.userName}}
+   * 
+   */
+  userName: string; // Amina.Langosh49
+
+  /**
+   * @timestamp past
```
Comment (Cryrivers) [inline README.md]: this syntax has been changed. i'll update it later.

File: README.md [hunk 3]
```diff
@@ -22,28 +87,41 @@ Contents
 npm install --save-dev @manta-style/cli
 ```
 
-You could also install it globally, which adds a command line tool `ms` to your system.
+To install globally
```
Comment (Cryrivers) [inline README.md]: according to the discussion this afternoon, we might not support global install at the moment.

File: README.md [hunk 4]
```diff
@@ -22,28 +87,41 @@ Contents
...
 
-Manta Style needs plugins to support different file types and generate mock data.
+This adds a command line tool `ms` to your system.
 
-The example of Quick Start below is using TypeScript. So first you might want to install TypeScript support to Manta Style.
+### Builders
+
+Manta Style supports generation of mock data from different typing systems via binding builders.
+
+In [Quick Start](#quick-start) we show an example using TypeScript. You will need to [install TypeScript](https://www.typescriptlang.org/index.html#download-links) _and_ add the Manta Style TypeScript builder:
```
Comment (Cryrivers) [inline README.md]: by right you don't need to install `typescript` in your project. if it's not the case, i'll create a bug ticket to myself.

File: README.md [hunk 5]
```diff
@@ -22,28 +87,41 @@ Contents
...
 
-```sh
-npm install --save-dev @manta-style/plugin-mock-example @manta-style/plugin-mock-faker
-```
+Manta Style supports a handful of data generating features via plugins.
+One major example is `@manta-style/plugin-mock-faker`. This is a plugin that enables you to generate fake datas that look real.
 
 You could check [Plugins](#plugins) for the usages of official plugins. You could make our own plugins as well.
 
+<!-- TODO: add support for creating plugins? -->
```
Comment (Cryrivers) [inline README.md]: yeah, i think @tanhauhau and me can provide more information for creating plugins.

**Comment by Cryrivers** [comment]
iirc I set external config. So I just need to remove typescript from peerDependency

File: README.md (hunk 2 above)
Comment (Cryrivers) [reply README.md]: oh it's not in this repo so i cannot idea. new syntax is `@faker date.past`

Comment (wgao19) [reply README.md]: I think you can edit in your main repo, so before I merge I will resolve conflict and use yours.

Comment (Cryrivers) [reply README.md]: try rebasing.

File: README.md [hunk 6]
```diff
@@ -69,7 +69,7 @@ type WithResponse<T> = WithResponseSuccess<T> | WithResponseFailure;
 Manta Style will generate either the error case or the normal case according to your declarations.
 Already have your list of error conditions and messages well-defined with your back end? You are in the luck! Manta Style will generate those cases for you just like in real world.
 
-This also gives us one more motivation to fine-tune the typing to our codebase.
```
Comment (wgao19) [inline README.md]: This is a verb and should use present form perhaps?

File: README.md [hunk 7]
```diff
@@ -69,7 +69,7 @@ type WithResponse<T> = WithResponseSuccess<T> | WithResponseFailure;
...
+This also gives us one more motivation to fine-tuned the typing to our codebase.
```
Comment (wgao19) [inline README.md]: This is a verb and should use present form perhaps?

File: README.md (hunk 6 above)
Comment (Cryrivers) [reply README.md]: oh i was being blind..sorry, will change it back.

File: README.md (hunk 7 above)
Comment (Cryrivers) [reply README.md]: sorry i was being blind 😂, i just read it as `to have fine-tuned typings in our codebase`.

will change it back.

**Comment by wgao19** [comment]
Hey @Cryrivers would you like to try setting up gitbook (or github pages) to host our `documentation` directory so we can have a generated docsite? Assuming this, I'm thinking of moving sections _Installation_ and _Quick Start_ to under documentations directory. Suggesting this structure because most major frameworks such as React and Redux are following this fashion.

**Comment by Cryrivers** [comment]
Wow that would be good. Please try it if you want. Can also let me set up some git hooks if you want to.

**Comment by Cryrivers** [comment]
@wgao19 thanks!!! for syntax part, i guess we should put them into builders folder. As we need to write syntax documentation for `TypeScript`and `Flow`. For TypeScript syntax, if you are interested, I can show you face 2 face! we can merge it after review for plugins part!

File: documentation/Plugins.md [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+
+#### Usage
+
+For strings, our Mock Faker users Faker.js's `Faker.fake()`.
+
+Mock Faker is transparent to Faker.js's. You may use any available Faker.js's API.
+
+You can check out Faker.js's `Faker.fake()` API [here](http://marak.github.io/faker.js/#toc7__anchor)
+
+For numbers, `plugin-mock-faker` calls `faker.method.path()`.
```
Comment (Cryrivers) [inline documentation/Plugins.md]: it calls the actual method in `faker`. for example, if you provide `date.past`, it calls `faker.date.past`.

**Comment by Cryrivers** [comment]
wow i just read the getting started part, it's super super impressive! things i can think of to show off are conditional types and custom type `Query` (a type for reading URL query from user input). and another thing worth mentioning is that user can write their own custom type and plugin.

For example:
```typescript
import { ShopeeAvatar } from '@shopee_common/manta-style-support';

type UserProfile<UserId extends number> = {
  userid: UserId,
  avatar: ShopeeAvatar<UserId>
}
```

Custom type `ShopeeAvatar` could get the url of shopee user profile picture. I had a chat with @tanhauhau last week, thinking `Manta Style` could really be the real server doing the real business logic with custom types and plugins, because for example, we could have a custom type to do MySQL queries.

Maybe Manta Style could be a nice way to implement a simple server in semi-declarative manners, which is super exciting to me.

**Pull request merged**